lib/Mojo/WebSocketProxy/Dispatcher.pm

```diff
@@ -69,20 +69,41 @@ sub open_connection {
         # Incoming data will be JSON-formatted text, as a Unicode string.
         # We normalize the entire string before decoding.
         my $decoded = eval { Encode::decode_utf8($msg) } or do {
-            stats_inc("websocket_proxy.utf8_decoding.failure", {tags => ['error_code:1007']});
-            $c->finish(1007 => 'Malformed UTF-8 data');
+            $config->{on_error}({
+                error    => 'Error Processing Request',
+                details  => { 
+                    connection      => $c,
+                    error_code      => 'INVALID_UTF8',
+                    reason          => 'Malformed UTF-8 data',
+                    request_body    => $msg,
+                },
+                }) if exists $config->{on_error};
             return;
         };
 
         my $normalized_msg = eval { Unicode::Normalize::NFC($decoded) } or do {
-            stats_inc("websocket_proxy.unicode_normalisation.failure", {tags => ['error_code:1007']});
-            $c->finish(1007 => 'Malformed Unicode data');
+            $config->{on_error}({
```

Comment:
Can you use `->emit` instead? Mojolicious already has some basic event handling built in.

See https://github.com/binary-com/perl-Mojo-WebSocketProxy/blob/master/lib/Mojo/WebSocketProxy/Parser.pm#L52 for an example.

Reply:
changes are made based on the comment